// webflow-api/api/resources/sites/resources/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.forms = exports.scripts = exports.activityLogs = exports.robotsTxt = exports.plans = exports.redirects = exports.comments = exports.wellKnown = void 0;
exports.wellKnown = __importStar(require("./wellKnown"));
__exportStar(require("./wellKnown/types"), exports);
exports.comments = __importStar(require("./comments"));
__exportStar(require("./comments/types"), exports);
exports.redirects = __importStar(require("./redirects"));
exports.plans = __importStar(require("./plans"));
exports.robotsTxt = __importStar(require("./robotsTxt"));
exports.activityLogs = __importStar(require("./activityLogs"));
exports.scripts = __importStar(require("./scripts"));
exports.forms = __importStar(require("./forms"));
__exportStar(require("./wellKnown/client/requests"), exports);
__exportStar(require("./activityLogs/client/requests"), exports);
__exportStar(require("./comments/client/requests"), exports);
__exportStar(require("./scripts/client/requests"), exports);
__exportStar(require("./forms/client/requests"), exports);
