// webflow-api/serialization/resources/users/types/UserAccountDeletedPayloadPayload.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserAccountDeletedPayloadPayload = void 0;
const core = __importStar(require("../../../../core"));
const UserAccountDeletedPayloadPayloadData_1 = require("./UserAccountDeletedPayloadPayloadData");
const UserStatus_1 = require("../../../types/UserStatus");
const UserAccessGroupsItem_1 = require("../../../types/UserAccessGroupsItem");
exports.UserAccountDeletedPayloadPayload = core.serialization.object({
    data: UserAccountDeletedPayloadPayloadData_1.UserAccountDeletedPayloadPayloadData.optional(),
    id: core.serialization.string().optional(),
    isEmailVerified: core.serialization.boolean().optional(),
    lastUpdated: core.serialization.date().optional(),
    invitedOn: core.serialization.date().optional(),
    createdOn: core.serialization.date().optional(),
    lastLogin: core.serialization.date().optional(),
    status: UserStatus_1.UserStatus.optional(),
    accessGroups: core.serialization.list(UserAccessGroupsItem_1.UserAccessGroupsItem).optional(),
});
