// webflow-api/dist/serialization/types/Form.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Form = void 0;
const core = __importStar(require("../../core"));
const FormField_1 = require("./FormField");
const FormResponseSettings_1 = require("./FormResponseSettings");
exports.Form = core.serialization.object({
    displayName: core.serialization.string().optional(),
    createdOn: core.serialization.date().optional(),
    lastUpdated: core.serialization.date().optional(),
    fields: FormField_1.FormField.optional(),
    responseSettings: FormResponseSettings_1.FormResponseSettings.optional(),
    id: core.serialization.string().optional(),
    siteId: core.serialization.string().optional(),
    siteDomainId: core.serialization.string().optional(),
    pageId: core.serialization.string().optional(),
    pageName: core.serialization.string().optional(),
    formElementId: core.serialization.string().optional(),
    workspaceId: core.serialization.string().optional(),
});
