// webflow-api/dist/serialization/types/SiteActivityLogItem.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SiteActivityLogItem = void 0;
const core = __importStar(require("../../core"));
const SiteActivityLogItemEvent_1 = require("./SiteActivityLogItemEvent");
const SiteActivityLogItemResourceOperation_1 = require("./SiteActivityLogItemResourceOperation");
const SiteActivityLogItemUser_1 = require("./SiteActivityLogItemUser");
exports.SiteActivityLogItem = core.serialization.object({
    id: core.serialization.string().optional(),
    createdOn: core.serialization.date().optional(),
    lastUpdated: core.serialization.date().optional(),
    event: SiteActivityLogItemEvent_1.SiteActivityLogItemEvent.optional(),
    resourceOperation: SiteActivityLogItemResourceOperation_1.SiteActivityLogItemResourceOperation.optional(),
    user: SiteActivityLogItemUser_1.SiteActivityLogItemUser.optional(),
    resourceId: core.serialization.string().optional(),
    resourceName: core.serialization.string().optional(),
    newValue: core.serialization.string().optional(),
    previousValue: core.serialization.string().optional(),
    payload: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
});
