// webflow-api/dist/serialization/types/Site.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Site = void 0;
const core = __importStar(require("../../core"));
const Domain_1 = require("./Domain");
const Locales_1 = require("./Locales");
const SiteDataCollectionType_1 = require("./SiteDataCollectionType");
exports.Site = core.serialization.object({
    id: core.serialization.string(),
    workspaceId: core.serialization.string().optional(),
    createdOn: core.serialization.date().optional(),
    displayName: core.serialization.string().optional(),
    shortName: core.serialization.string().optional(),
    lastPublished: core.serialization.date().optional(),
    lastUpdated: core.serialization.date().optional(),
    previewUrl: core.serialization.string().optional(),
    timeZone: core.serialization.string().optional(),
    parentFolderId: core.serialization.string().optional(),
    customDomains: core.serialization.list(Domain_1.Domain).optional(),
    locales: Locales_1.Locales.optional(),
    dataCollectionEnabled: core.serialization.boolean().optional(),
    dataCollectionType: SiteDataCollectionType_1.SiteDataCollectionType.optional(),
});
