// webflow-api/dist/serialization/types/WorkspaceAuditLogItem.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceAuditLogItem = void 0;
const WorkspaceAuditLogItemActor_1 = require("./WorkspaceAuditLogItemActor");
const WorkspaceAuditLogItemWorkspace_1 = require("./WorkspaceAuditLogItemWorkspace");
const core = __importStar(require("../../core"));
const UserAccessAuditLogItem_1 = require("./UserAccessAuditLogItem");
const CustomRoleAuditLogItem_1 = require("./CustomRoleAuditLogItem");
const WorkspaceMembershipAuditLogItem_1 = require("./WorkspaceMembershipAuditLogItem");
const SiteMembershipAuditLogItem_1 = require("./SiteMembershipAuditLogItem");
const WorkspaceInvitationAuditLogItem_1 = require("./WorkspaceInvitationAuditLogItem");
const _Base = core.serialization.object({
    timestamp: core.serialization.date().optional(),
    actor: WorkspaceAuditLogItemActor_1.WorkspaceAuditLogItemActor.optional(),
    workspace: WorkspaceAuditLogItemWorkspace_1.WorkspaceAuditLogItemWorkspace.optional(),
});
exports.WorkspaceAuditLogItem = core.serialization
    .union("eventType", {
    user_access: UserAccessAuditLogItem_1.UserAccessAuditLogItem.extend(_Base),
    custom_role: CustomRoleAuditLogItem_1.CustomRoleAuditLogItem.extend(_Base),
    workspace_membership: WorkspaceMembershipAuditLogItem_1.WorkspaceMembershipAuditLogItem.extend(_Base),
    site_membership: SiteMembershipAuditLogItem_1.SiteMembershipAuditLogItem.extend(_Base),
    workspace_invitation: WorkspaceInvitationAuditLogItem_1.WorkspaceInvitationAuditLogItem.extend(_Base),
})
    .transform({
    transform: (value) => value,
    untransform: (value) => value,
});
