// webflow-api/serialization/types/Node.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Node = void 0;
const core = __importStar(require("../../core"));
const TextNode_1 = require("./TextNode");
const ImageNode_1 = require("./ImageNode");
const ComponentNode_1 = require("./ComponentNode");
const TextInputNode_1 = require("./TextInputNode");
const SelectNode_1 = require("./SelectNode");
const SubmitButtonNode_1 = require("./SubmitButtonNode");
const SearchButtonNode_1 = require("./SearchButtonNode");
exports.Node = core.serialization
    .union("type", {
    text: TextNode_1.TextNode,
    image: ImageNode_1.ImageNode,
    "component-instance": ComponentNode_1.ComponentNode,
    "text-input": TextInputNode_1.TextInputNode,
    select: SelectNode_1.SelectNode,
    "submit-button": SubmitButtonNode_1.SubmitButtonNode,
    "search-button": SearchButtonNode_1.SearchButtonNode,
})
    .transform({
    transform: (value) => value,
    untransform: (value) => value,
});
