// webflow-api/serialization/types/SkuFieldData.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkuFieldData = void 0;
const core = __importStar(require("../../core"));
const SkuValueList_1 = require("./SkuValueList");
const SkuFieldDataPrice_1 = require("./SkuFieldDataPrice");
const SkuFieldDataCompareAtPrice_1 = require("./SkuFieldDataCompareAtPrice");
const SkuFieldDataEcSkuBillingMethod_1 = require("./SkuFieldDataEcSkuBillingMethod");
const SkuFieldDataEcSkuSubscriptionPlan_1 = require("./SkuFieldDataEcSkuSubscriptionPlan");
const SkuPropertyList_1 = require("./SkuPropertyList");
exports.SkuFieldData = core.serialization.object({
    skuValues: core.serialization.property("sku-values", SkuValueList_1.SkuValueList.optional()),
    name: core.serialization.string(),
    slug: core.serialization.string(),
    price: SkuFieldDataPrice_1.SkuFieldDataPrice,
    compareAtPrice: core.serialization.property("compare-at-price", SkuFieldDataCompareAtPrice_1.SkuFieldDataCompareAtPrice.optional()),
    ecSkuBillingMethod: core.serialization.property("ec-sku-billing-method", SkuFieldDataEcSkuBillingMethod_1.SkuFieldDataEcSkuBillingMethod.optional()),
    ecSkuSubscriptionPlan: core.serialization.property("ec-sku-subscription-plan", SkuFieldDataEcSkuSubscriptionPlan_1.SkuFieldDataEcSkuSubscriptionPlan.optional()),
    mainImage: core.serialization.property("main-image", core.serialization.string().optional()),
    sku: core.serialization.string().optional(),
    skuProperties: core.serialization.property("sku-properties", core.serialization.list(SkuPropertyList_1.SkuPropertyList).optional()),
});
