// webflow-api/api/resources/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ecommerce = exports.forms = exports.webhooks = exports.assets = exports.scripts = exports.token = exports.workspaces = exports.collections = exports.inventory = exports.orders = exports.products = exports.accessGroups = exports.users = exports.components = exports.pages = exports.sites = void 0;
exports.sites = __importStar(require("./sites"));
__exportStar(require("./sites/types"), exports);
exports.pages = __importStar(require("./pages"));
__exportStar(require("./pages/types"), exports);
exports.components = __importStar(require("./components"));
__exportStar(require("./components/types"), exports);
exports.users = __importStar(require("./users"));
__exportStar(require("./users/types"), exports);
exports.accessGroups = __importStar(require("./accessGroups"));
__exportStar(require("./accessGroups/types"), exports);
exports.products = __importStar(require("./products"));
__exportStar(require("./products/types"), exports);
exports.orders = __importStar(require("./orders"));
__exportStar(require("./orders/types"), exports);
exports.inventory = __importStar(require("./inventory"));
__exportStar(require("./inventory/types"), exports);
exports.collections = __importStar(require("./collections"));
exports.workspaces = __importStar(require("./workspaces"));
exports.token = __importStar(require("./token"));
exports.scripts = __importStar(require("./scripts"));
exports.assets = __importStar(require("./assets"));
exports.webhooks = __importStar(require("./webhooks"));
exports.forms = __importStar(require("./forms"));
exports.ecommerce = __importStar(require("./ecommerce"));
__exportStar(require("./sites/client/requests"), exports);
__exportStar(require("./collections/client/requests"), exports);
__exportStar(require("./pages/client/requests"), exports);
__exportStar(require("./components/client/requests"), exports);
__exportStar(require("./scripts/client/requests"), exports);
__exportStar(require("./assets/client/requests"), exports);
__exportStar(require("./forms/client/requests"), exports);
__exportStar(require("./users/client/requests"), exports);
__exportStar(require("./accessGroups/client/requests"), exports);
__exportStar(require("./products/client/requests"), exports);
__exportStar(require("./orders/client/requests"), exports);
__exportStar(require("./inventory/client/requests"), exports);
