// webflow-api/dist/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebflowClient = void 0;
const core = __importStar(require("./core"));
const headers_js_1 = require("./core/headers.js");
const Client_1 = require("./api/resources/token/client/Client");
const Client_2 = require("./api/resources/sites/client/Client");
const Client_3 = require("./api/resources/collections/client/Client");
const Client_4 = require("./api/resources/pages/client/Client");
const Client_5 = require("./api/resources/components/client/Client");
const Client_6 = require("./api/resources/scripts/client/Client");
const Client_7 = require("./api/resources/assets/client/Client");
const Client_8 = require("./api/resources/webhooks/client/Client");
const Client_9 = require("./api/resources/forms/client/Client");
const Client_10 = require("./api/resources/users/client/Client");
const Client_11 = require("./api/resources/accessGroups/client/Client");
const Client_12 = require("./api/resources/products/client/Client");
const Client_13 = require("./api/resources/orders/client/Client");
const Client_14 = require("./api/resources/inventory/client/Client");
const Client_15 = require("./api/resources/ecommerce/client/Client");
const Client_16 = require("./api/resources/workspaces/client/Client");
class WebflowClient {
    constructor(_options = {}) {
        this._options = Object.assign(Object.assign({}, _options), { headers: (0, headers_js_1.mergeHeaders)({
                "X-Fern-Language": "JavaScript",
                "X-Fern-SDK-Name": "webflow-api",
                "X-Fern-SDK-Version": "3.2.2",
                "User-Agent": "webflow-api/3.2.2",
                "X-Fern-Runtime": core.RUNTIME.type,
                "X-Fern-Runtime-Version": core.RUNTIME.version,
            }, _options === null || _options === void 0 ? void 0 : _options.headers) });
    }
    get token() {
        var _a;
        return ((_a = this._token) !== null && _a !== void 0 ? _a : (this._token = new Client_1.Token(this._options)));
    }
    get sites() {
        var _a;
        return ((_a = this._sites) !== null && _a !== void 0 ? _a : (this._sites = new Client_2.Sites(this._options)));
    }
    get collections() {
        var _a;
        return ((_a = this._collections) !== null && _a !== void 0 ? _a : (this._collections = new Client_3.Collections(this._options)));
    }
    get pages() {
        var _a;
        return ((_a = this._pages) !== null && _a !== void 0 ? _a : (this._pages = new Client_4.Pages(this._options)));
    }
    get components() {
        var _a;
        return ((_a = this._components) !== null && _a !== void 0 ? _a : (this._components = new Client_5.Components(this._options)));
    }
    get scripts() {
        var _a;
        return ((_a = this._scripts) !== null && _a !== void 0 ? _a : (this._scripts = new Client_6.Scripts(this._options)));
    }
    get assets() {
        var _a;
        return ((_a = this._assets) !== null && _a !== void 0 ? _a : (this._assets = new Client_7.Assets(this._options)));
    }
    get webhooks() {
        var _a;
        return ((_a = this._webhooks) !== null && _a !== void 0 ? _a : (this._webhooks = new Client_8.Webhooks(this._options)));
    }
    get forms() {
        var _a;
        return ((_a = this._forms) !== null && _a !== void 0 ? _a : (this._forms = new Client_9.Forms(this._options)));
    }
    get users() {
        var _a;
        return ((_a = this._users) !== null && _a !== void 0 ? _a : (this._users = new Client_10.Users(this._options)));
    }
    get accessGroups() {
        var _a;
        return ((_a = this._accessGroups) !== null && _a !== void 0 ? _a : (this._accessGroups = new Client_11.AccessGroups(this._options)));
    }
    get products() {
        var _a;
        return ((_a = this._products) !== null && _a !== void 0 ? _a : (this._products = new Client_12.Products(this._options)));
    }
    get orders() {
        var _a;
        return ((_a = this._orders) !== null && _a !== void 0 ? _a : (this._orders = new Client_13.Orders(this._options)));
    }
    get inventory() {
        var _a;
        return ((_a = this._inventory) !== null && _a !== void 0 ? _a : (this._inventory = new Client_14.Inventory(this._options)));
    }
    get ecommerce() {
        var _a;
        return ((_a = this._ecommerce) !== null && _a !== void 0 ? _a : (this._ecommerce = new Client_15.Ecommerce(this._options)));
    }
    get workspaces() {
        var _a;
        return ((_a = this._workspaces) !== null && _a !== void 0 ? _a : (this._workspaces = new Client_16.Workspaces(this._options)));
    }
}
exports.WebflowClient = WebflowClient;
